server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+import static io.crate.sql.tree.JoinType.CROSS;
+import static io.crate.sql.tree.JoinType.INNER;
+import static io.crate.sql.tree.JoinType.LEFT;
+import static io.crate.sql.tree.JoinType.RIGHT;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+
+import io.crate.analyze.relations.QuerySplitter;
+import io.crate.common.collections.Sets;
+import io.crate.expression.operator.AndOperator;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.TransactionContext;
+import io.crate.planner.consumer.RelationNameCollector;
+import io.crate.planner.operators.Filter;
+import io.crate.planner.operators.JoinPlan;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.costs.PlanStats;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.sql.tree.JoinType;
+
+/**
+ * This rule extracts constant join conditions from inner/outer/cross joins and places them as
+ * a filter on top of the join. Then other rules can process the filter further.
+ * <pre>
+ * Join[INNER | ((x = y) AND (y > 1))]
+ *   ├ Collect[doc.t1 | [x] | true]
+ *   └ Collect[doc.t2 | [y] | true]
+ * </pre>
+ * becomes:
+ * <pre>
+ * Filter[(y > 1)]
+ * └ Join[INNER | (x = y)]
+ *     ├ Collect[doc.t1 | [x] | true]
+ *     └ Collect[doc.t2 | [y] | true]
+ * </pre>
+ */
+public class ExtractConstantJoinConditionsToFilter implements Rule<JoinPlan> {
+
+    private static final Set<JoinType> SUPPORTED_JOIN_TYPES = EnumSet.of(INNER, LEFT, RIGHT, CROSS);
+    private final Pattern<JoinPlan> pattern;
+
+    public ExtractConstantJoinConditionsToFilter() {
+        this.pattern = typeOf(JoinPlan.class)
+            .with(join ->
+                SUPPORTED_JOIN_TYPES.contains(join.joinType()) &&
+                !join.joinConditionsExtractedToFilters() &&
+                join.joinCondition() != null &&
+                hasConstantJoinConditions(join.joinCondition())
+            );
+    }
+
+    private static boolean hasConstantJoinConditions(Symbol joinCondition) {
```

Comment:
I'd remove this method and the check in the pattern matching, to avoid calling the `QuerySplitter.split()` twice. I think it's fine to just bail out early in `apply()` or "cache" the outcome of the splitter in a variable.

Reply:
This is on purpose in the pattern matching, to make sure this rule only matches if there is a constant join condition present. If i put this into the apply it will show up on `explain verbose` without doing anything.

Reply:
Makes sense, thx for explaining! Consider if there is a solution to not do the query splitting twice then.